pkg/channel/distributed/dispatcher/dispatcher/handler.go

```diff
@@ -147,54 +146,3 @@ func (h *Handler) consumeMessage(context context.Context, consumerMessage *saram
 	_, dispatchError := h.MessageDispatcher.DispatchMessageWithRetries(ctx, message, nil, destinationURL, replyURL, deadLetterURL, retryConfig)
 	return dispatchError
 }
-
-//
-// Custom Implementation Of RetryConfig.CheckRetry To Determine Whether To Retry Based On Response
-//
-// Note - Returning true indicates a retry should occur.  Returning an error will result in that
-//        error being returned instead of any errors from the Request.
-//
-func (h *Handler) checkRetry(_ context.Context, response *http.Response, err error) (bool, error) {
-
-	// Retry Any Nil HTTP Response
-	if response == nil {
-		h.Logger.Info("Unable To Check Retry State On Nil Response - Retrying")
-		return true, nil
-	}
-
-	// Retry Any Errors
-	if err != nil {
-		h.Logger.Warn("Received Response Error - Retrying", zap.Error(err))
-		return true, nil
-	}
-
-	// Extract The StatusCode From The Response & Add To Logger
-	statusCode := response.StatusCode
-	logger := h.Logger.With(zap.Int("StatusCode", statusCode))
-
-	// Note - Normally we would NOT want to retry 4xx responses, BUT there are a few
-	//        known areas of knative-eventing that return codes in this range which
-	//        require retries.  Reasons for particular codes are as follows:
-	//
-	// 404  Although we would ideally not want to retry a permanent "Not Found"
-	//      response, a 404 can be returned when a pod is in the process of becoming
-	//      ready, so a retry can be a useful thing in this situation.
-	// 409  Returned by the E2E tests, so we must retry when "Conflict" is received, or the
-	//      tests will fail (see knative.dev/eventing/test/lib/recordevents/receiver/receiver.go)
-	// 429  Since retry typically involves a delay (usually an exponential backoff),
-	//      retrying after receiving a "Too Many Requests" response is useful.
-
-	if statusCode >= 500 || statusCode == 404 || statusCode == 429 || statusCode == 409 {
-		logger.Warn("Failed To Send Message To Subscriber Service - Retrying")
-		return true, nil
-	} else if statusCode >= 300 && statusCode <= 399 {
-		logger.Warn("Failed To Send Message To Subscriber Service - Not Retrying")
-		return false, nil
-	} else if statusCode == -1 {
-		logger.Warn("No StatusCode Detected In Error - Retrying")
-		return true, nil
-	}
-
-	// Do Not Retry 1XX, 2XX, & Most 4XX StatusCode Responses
-	return false, nil
-}
```

Comment:
So... what's the rationale for moving this function into eventing ?   I had always viewed this as the distributed KafkaChannel's opinionated view of "when" to retry.  The RetryConfig exposes this hook for implementations to customize the retry decision.

Are you planning on applying the same logic to the consolidated KafkaChannel or using it somewhere else?  Is this part of the larger plan to expose configuration in the Delivery block for the retry decision?  I'm ok with it living in eventing, as long as we can make changes to it in the interim before we have this full configuration capability.

Reply:
> Are you planning on applying the same logic to the consolidated KafkaChannel

yes, that's the plan

Reply:
and yes - we can make changes - sure

Reply:
I like this implementation much better - and would even want to deprecate the old one, in the long run, hence the motivation for moving out

Reply:
Ok sounds good - I just saw same discussion in issue #398 - sorry for duplicating here - I was out for a couple days and wasn't paying attention (obviously : )